### PR TITLE
rgw: make rgw-restore-bucket-index more robust

### PR DESCRIPTION
**Pull request by ivancich**
This experimental tool write a series of temporary files, the combined size of which is roughly proportional to some combination of the number of objects in the bucket and their respective sizes. Previously the script stored all temporary files in /tmp, which may not be sufficiently large in many cases.

There are now two enhancements. First, using the "-t" command-line argument, the directory for temporary files can be specified. Second, if the script consumes all available space on the file system containing the temporary directory, an error message to that effect is displayed, so the user understands what corrective action to take.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by ivancich** [comment]
This can be tested by creating a small temporary files system and using it for temporary files. Such as:

`sudo mount -o size=12K -t tmpfs none /mnt/tmpfs`

**Comment by ivancich** [comment]
@linuxkidd: I'd appreciate any corrections or suggestions for improvement.

File: src/rgw/rgw-restore-bucket-index [hunk 1]
```diff
@@ -23,35 +23,48 @@
...
   fi
 }
 
+test_temp_space() {
+    # if there is a command-line argument, assume it's exit status
+    # use exit status if provided, otherwise ignore
+    test_space=0
+    if [ $# -eq 0 ] ;then
+	test_space=1
+    elif [ "$1" -ne 0 ] ;then
```
Comment (linuxkidd) [inline src/rgw/rgw-restore-bucket-index:53]: I'm not sure why we'd want to call `test_temp_space()` with no intention of actually testing the space. :)

**Comment by linuxkidd** [comment]
Aside from the `test_temp_space` comment, the update looks great to me!

File: src/rgw/rgw-restore-bucket-index (hunk 1 above)
Comment (ivancich) [reply src/rgw/rgw-restore-bucket-index:53]: My original intent was to only test space when the command-line that was writing output there exited with an error code (i.e., non-zero). But then I had to relax that once I realized that there were problems if the exit status was generated by an "eval" or possibly whe a more complex control structure was used (output comes from an entire while loop).

But you're right; let's not even test the exit status and just check disk space.

Thanks, @linuxkidd .

**Review by linuxkidd — APPROVED** [review]

**Comment by ivancich** [comment]
Teuthology results -- no regressions:
* https://pulpito.ceph.com/ivancich-2024-03-06_17:23:45-rgw-wip-eric-testing-1-distro-default-smithi/
* https://pulpito.ceph.com/ivancich-2024-03-06_17:24:06-rgw-wip-eric-testing-1-distro-default-smithi/

**Pull request merged**